misc/man/rrr_logd.1
.Dd 2025-02-15
.TH RRR_LOGD 1
.SH NAME
Read Route Record is a dynamic module-based message handling program
for collecting, buffering, processing and storing messages and measurements.
.PP
.B rrr_logd,
the RRR log daemon, receives log messages on a socket from either
.B rrr_post
or the
.B rrr
main daemon. Multiple processes may connect to the log daemon socket,
and the log messages are printed to standard output.
.PP
The log daemon may help in situations where standard output from multiple processes gets interleaved. 
.SH SYNOPSIS
.B rrr_logd [OPTIONS] [WRAPPER AND ARGS]
.Dl [-s|--socket[=]SOCKET]
.Dl [-f|--file-descriptor[=]FILE DESCRIPTOR]
.Dl [-p|--persist]
.Dl [-q|--quiet]
.Dl [-a|--add-newline]
.Dl [-m|--message-only]
.Dl [-j|--json]
.Dl [-u|--unbuffered]
.Dl [-l|--loglevel-translation]
.Dl [-e|--environment-file[=]ENVIRONMENT FILE]
.Dl [-d|--debuglevel[=]DEBUG FLAGS]
.Dl [-D|--debuglevel-on-exit[=]DEBUG FLAGS]
.Dl [-h|--help]
.Dl [-v|--version]

.SH DESCRIPTION
The log daemon may receive both native RRR log messages and RRR array messages with specific fields set.

If array messages are used, the following fields are expected to be present of which most are optional:

If a wrapper process is specified, each line will be written to this process, and output is read back.
This allows for modifying logs before printed out.

.Bl -tag -width -indent

.It log_level_translated
RFC 5424 loglevel, small integer. If not given, 7 is used.

.It log_flags
If given and set to 1, this indicates that the log message is in JSON format.

.It log_prefix
The log prefix. If not given, "rrr_logd" will be used.

.It log_message
The actual log message. This value is mandatory.

.It log_file
.It log_line
The file name and line from where the log message originates, string and integer.
If either is not given, values from the
.B rrr_logd
code will be used.

.El

.SH OPTIONS
.IP -s|--socket[=SOCKET]
The filename of the socket on which to listen for incoming connections from other RRR processes.

.IP -f|--file-descriptor[=FILE DESCRIPTOR]
If
.B rrr_logd
is started by for instance
.Xr execve(2)
after a fork, a numeric file handle may be given for a socket on which the parent process sends messages.

.IP -p|--persist
If not set, the process will terminate once the last user has disconnected.

.IP -q|--quiet
Don't print any messages received from other processes. Used for debugging.

.IP -n|--add-newline
Add a newline (ASCII decimal 10) to any message received not already ending with one.

.IP -m|--message-only
Print only the message text and not log level and prefix.

.IP -j|--json
If given, all data will be printed in JSON format.
The fields present in the JSON output correspond to the fields of array log messages (see above).

If the JSON flag is set in
.B log_flags
or in the flags of native RRR log messages, the JSON in those log messages will be merged with these fields.
If any field names collide, existing fields from the JSON will be overwritten silently.

If
.B --message-only
is set, ony the field
.B log_message
will be present in the JSON.

JSON data will be printed with one log object per line, with lines separated by newline, regardless of the
.B --add-newline
parameter.

.IP -u|--unbuffered
Make standard output unbuffered.

.IP -l|--loglevel-translation
Translate any RRR native log level in received messages to RFC 5424 format.
This will only affect messages coming from the main RRR daemon.
Messages from
.B rrr_post
must always have RFC 5424 format.

.IP -d|--debuglevel[=]FLAGS
Debuglevel to use. Will only affect log messages internal to
.B rrr_logd.

.IP -e|--environment-file
See
.Xr rrr(1)
for usage.

.SH SIGNALS
.IP SIGURS1
Terminal process nicely.
.IP SIGINT/CTRL+C
Terminal process nicely.
.IP SIGTERM
Terminal process nicely.
.IP SIGPIPE
Ignored (an error message is printed).
.PP

.SH RETURN VALUE
.B rrr_post
returns 0 on success and 1 on failure.

.SH DEBUG FLAGS
Refer to
.Xr rrr(1)
for possible debug flags.

.SH SEE ALSO
.Xr rrr(1),
.Xr rrr_post(1)


debian/changelog
rrr (1.36-15) bookworm; urgency=medium

  * RUNNING DEVELOPMENT RELEASE

  * Features:
    * New methods on JS Message object: data_as_object and data_as_utf8
    * Support for Node 23
    * Support for Debian Trixie
    * Log daemon may start a wrapper process for further processing of
      messages prior to them being printed.
    * Configurable sub instances in cmodules
    * Native JSON logging from log daemon

  * Fixes:
    * BAIL in method and route definitions not working correctly, caused
      program exit when triggered
    * FALSE value in method and route definitions not working, TRUE was pushed
      instead
    * Missing check for overshoot and undershoot while fading in ArtNet
    * Incorrect counting of FIFO buffer entries in certain situations, could
      possibly result in buffers over time being reported as full when they
      were not. The problem only affected counting of retained messages in MQTT
      broker.
    * Misc. compilation problems using libc++ on FreeBSD, which is required
      then V8 is used on this platform
    * MQTT client not always reconnecting as configured but restarted
    * JS incorrectly loading the same module multiple times if included
      from different scripts

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 24 Mar 2025 12:04:31 +0100

rrr (1.35-1) bookworm; urgency=medium

  * Features:
    * Configuration option in http client to trap any header field

  * Fixes:
    * Crashes in MQTT for malformed packets with tailing payload
    * Bad MQTT CONNECT packet assembling for V5 when will flag was not set
    * Log messages to stats engine  getting lost due to events delivered to
      incorrect event queue
    * Memory leak in Python3 where values added to array had incorrect
      reference count causing memory leaks
    * Broken parsing of leading spaces in istr and ustr types in arrays
    * Crash in array tree import for division overflow in conditions,
      emit soft error instead.

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 27 Nov 2024 18:32:12 +0100

rrr (1.33-1) bookworm; urgency=medium

  * Features:
    * Log socket daemon with ability to collect logs from multiple RRR
      processes

  * Improvements:
    * Remove all references to http2 when RRR is built without nghttp2, in
      particular avoid noisy notification caused by inconsistencies in the
      code
    * Speed up tests by running them in parallel
    * In httpclient, by default ignore all meta fields from httpserver which
      should usually not be added as request fields.
    * In httpclient, prevent using // endpoint when using endpoint from topic
      and url is /

  * Fixes:
    * Remove stale python directory from source causing build to fail in some
      cases
    * Misc. hangs and crashes when parsing invalid configuration files
    * Misc. crashes in HTTP handling

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 01 Nov 2024 07:02:09 +0100

rrr (1.31-5) bookworm; urgency=medium

  * Fixes:
    * Preflight headers not sendt for OPTIONS request
    * Move setproctitle() call to ensure that it cannot clear environment
      variables prior to them being accessed during initialization.

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 12 Aug 2024 13:05:28 +0200

rrr (1.31-1) bookworm; urgency=medium

  * Features:
    * New option http_server_decode_endpoint to decode the endpoint field
      when receiving full requests.
    * Support HTTP3 over QUIC in HTTP server and client.
    * Start RRR in single thread mode which is useful for debugging and
      on slower hardware.
    * Support for more ModBus functions
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New method rrr_array.has() in the Python module to check if an array tag
      exists.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New option http_server_topic_format to include more information about a
      request in the topic of messages generated by HTTP server module.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>

  * Improvements:
    * Implement polite close of connections during shutdown for HTTP1
      and HTTP2 in HTTP server.
    * Optimization in message database to avoid excessive chdir calls.

  * Fixes:
    * Broken TLS mode in MQTT client.
    * All settings used check not working correctly for JS module.

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 29 May 2024 08:43:01 +0200

rrr (1.29-2) bookworm; urgency=medium

  * Changes:
    * In route definition, the I operator is replaced with D. Existing programs
      must be updated to prevent warning from being printed.
    * Changed signatures for processor functions by adding a new last argument
      'method'. Existing scripts MAY BREAK and should be updated. Refer to
      the rrr.conf.5 manual page for correct function signatures.
    * Set process and thread names on OS-es where this is possible. This can
      assist when debugging for instance CPU usage.

  * Features:
    * Match generated messages in httpclient with their corresponding message
      triggering the query by using a report tag
    * Route messages leaving an instance to specific reader instances by topic
      or array tags
    * New parameter in httpclient to control timeout of low priority messages
      specifically
    * New array value "hdlc" for parsing HDLC messages
    * Ability in file module to write to files and devices among other new
      configuration parameters
    * New conversion type "hchar2str" in mangler module to convert individual
      bytes to strings
    * New module artnet for controlling DMX devices over ArtNet protocol
    * New OS object in javascript (operating-system) and ability to obtain
      hostname
    * New ModBus TCP module
    * New Lua module
    * New method on Message object in JS push_tag_object which serializes
      objects into JSON. Receiving JS instances will automatically de-
      serialize the JSON added to messages this way.

src/rrr_logd.c
/*

Read Route Record

Copyright (C) 2024-2025 Atle Solbakken atle@goliathdns.no

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>

#include "lib/log.h"
#include "lib/allocator.h"

#include "lib/messages/msg_head.h"
#include "lib/messages/msg_log.h"
#include "lib/messages/msg_msg_struct.h"
#include "lib/rrr_types.h"
#include "lib/socket/rrr_socket_constants.h"
#include "lib/util/macro_utils.h"
#include "main.h"
#include "../build_timestamp.h"
#include "lib/cmdlineparser/cmdline.h"
#include "lib/rrr_config.h"
#include "lib/rrr_strerror.h"
#include "lib/common.h"
#include "lib/version.h"
#include "lib/array.h"
#include "lib/fork.h"
#include "lib/socket/rrr_socket.h"
#include "lib/socket/rrr_socket_client.h"
#include "lib/event/event.h"
#include "lib/helpers/log_helper.h"

RRR_CONFIG_DEFINE_DEFAULT_LOG_PREFIX("rrr_logd");

static volatile int some_fork_has_stopped = 0;
static volatile int main_running = 1;
static volatile int sigusr2 = 0;

int rrr_signal_handler(int s, void *arg) {
	return rrr_signal_default_handler(&main_running, &sigusr2, s, arg);
}

static const struct cmd_arg_rule cmd_rules[] = {
        {CMD_ARG_FLAG_HAS_ARGUMENT,   's',    "socket",                "[-s|--socket[=]SOCKET]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,   'f',    "file-descriptor",       "[-f|--file-descriptor[=]FILE DESCRIPTOR]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'p',    "persist",               "[-p|--persist]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'q',    "quiet",                 "[-q|--quiet]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'n',    "add-newline",           "[-n|--add-newline]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'm',    "message-only",          "[-m|--message-only]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'j',    "json",                  "[-j|--json]"},
	{CMD_ARG_FLAG_NO_ARGUMENT,    'u',    "unbuffered",            "[-u|--unbuffered]"},
        {CMD_ARG_FLAG_NO_ARGUMENT,    'l',    "loglevel-translation",  "[-l|--loglevel-translation]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,   'e',    "environment-file",      "[-e|--environment-file[=]ENVIRONMENT FILE]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,   'd',    "debuglevel",            "[-d|--debuglevel[=]DEBUG FLAGS]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,   'D',    "debuglevel-on-exit",    "[-D|--debuglevel-on-exit[=]DEBUG FLAGS]"},
        {CMD_ARG_FLAG_NO_ARGUMENT,    'h',    "help",                  "[-h|--help]"},
        {CMD_ARG_FLAG_NO_ARGUMENT,    'v',    "version",               "[-v|--version]"},
	{CMD_ARG_FLAG_NO_FLAG_MULTI,  '\0',   "wrapper-and-args",      "[WRAPPER AND ARGS]"},
        {0,                           '\0',    NULL,                   NULL}
};

struct rrr_logd_data {
	const char *receive_socket;
	int receive_socket_fd;
	int receive_fd;
	int persist;
	int quiet;
	int add_newline;
	int message_only;
	int unbuffered;
	char **wrapper;
};

static void rrr_logd_cleanup_config (struct rrr_logd_data *data) {
	char *p;
	int i;
	if (data->wrapper != NULL) {
		for (i = 0, p = data->wrapper[0]; p != NULL; i++, p = data->wrapper[i]) {
			rrr_free(p);
		}
		rrr_free(data->wrapper);
	}
}

static int rrr_logd_parse_config (struct rrr_logd_data *data, struct cmd_data *cmd) {
	const char *receive_socket;
	const char *receive_fd_str;

	receive_socket = cmd_get_value(cmd, "socket", 0);
	if (cmd_get_value(cmd, "socket", 1) != NULL) {
		RRR_MSG_0("Argument 'socket' may not be specified multiple times\n");
		return 1;
	}

	receive_fd_str = cmd_get_value(cmd, "file-descriptor", 0);
	if (cmd_get_value(cmd, "file-descriptor", 1) != NULL) {
		RRR_MSG_0("Argument 'file-descriptor' may not be specified multiple times\n");
		return 1;
	}

	if (receive_socket == NULL && receive_fd_str == NULL) {
		RRR_MSG_0("Neither 'file-descriptor' nor 'socket' argument was specified\n");
		return 1;
	}

	if (receive_socket != NULL) {
		data->receive_socket = receive_socket;
	}

	if (receive_fd_str != NULL) {
		long fd_tmp;
		if (cmd_convert_integer_10(receive_fd_str, &fd_tmp) != 0) {
			RRR_MSG_0("Failed to convert 'file-descriptor' argument to integer\n");
			return 1;
		}
#if LONG_MAX > INT_MAX
		if (fd_tmp > INT_MAX) {
			RRR_MSG_0("Value for 'file-descriptor' too high\n");
			return 1;
		}
#endif
		if (fd_tmp < 2) {
			RRR_MSG_0("Value for 'file-descriptor' must be greater than 2\n");
			return 1;
		}

		data->receive_fd = rrr_int_from_slength_bug_const(fd_tmp);
	}

	if (cmd_exists(cmd, "persist", 0)) {
		data->persist = 1;
	}

	if (cmd_exists(cmd, "quiet", 0)) {
		data->quiet = 1;
	}

	if (cmd_exists(cmd, "add-newline", 0)) {
		data->add_newline = 1;
	}

	if (cmd_exists(cmd, "message-only", 0)) {
		data->message_only = 1;
	}

	if (cmd_exists(cmd, "unbuffered", 0)) {
		data->unbuffered = 1;
	}

	const char *wrapper_string;
	cmd_arg_count wrapper_count = 0;
	while ((wrapper_string = cmd_get_value(cmd, "wrapper-and-args", wrapper_count)) != NULL) {
		wrapper_count++;
	}

	if (wrapper_count > 0) {
		if ((data->wrapper = rrr_allocate_zero(sizeof(*data->wrapper) * (wrapper_count + 1))) == NULL) {
			RRR_MSG_0("Failed to allocate memory for arguments in %s\n", __func__);
			return 1;
		}

		for (cmd_arg_count i = 0; i < wrapper_count; i++) {
			const char *str = cmd_get_value(cmd, "wrapper-and-args", i);
			if ((data->wrapper[i] = rrr_strdup(str)) == NULL) {
				RRR_MSG_0("Failed to allocate memory for wrapper argument in %s\n");
				return 1;
			}
		}
	}

	/* Caller must call cleanup function upon fail */

	return 0;
}

static int rrr_logd_socket_setup(struct rrr_logd_data *data) {
	if (data->receive_socket == NULL)
		return 0;

	if (rrr_socket_unix_create_bind_and_listen (
			&data->receive_socket_fd,
			"rrr_logd",
			data->receive_socket,
			SOMAXCONN,
			1, /* Non-blocking */
			0, /* No mkstemp */
			0  /* No unlink if exists */
	) != 0) {
		return 1;
	}

	return 0;
}

struct rrr_logd_callback_data {
	struct rrr_logd_data *data;
	struct rrr_fork_handler *fork_handler;
	int fd_stdout_orig;
};

static void rrr_logd_stdout_restore (int fd_stdout_orig) {
	if (fd_stdout_orig) {
		if (dup2(fd_stdout_orig, STDOUT_FILENO) < 0) {
			// Further messages will not be printed, just crash to avoid confusion
			RRR_BUG("Failed to redirect output back to standard output, cannot handle this situation");
		}
		RRR_DBG_1("Redirected output back to standard output\n");
	}
}

static int rrr_logd_periodic (void *arg) {
	struct rrr_logd_callback_data *callback_data = arg;

	if (!main_running)
		return RRR_EVENT_EXIT;

	rrr_fork_handle_sigchld_and_notify_if_needed(callback_data->fork_handler, 0);

	if (some_fork_has_stopped) {
		rrr_logd_stdout_restore(callback_data->fd_stdout_orig);
		RRR_MSG_0("One or more forks has exited\n");
		return RRR_EVENT_EXIT;
	}

	return RRR_EVENT_OK;
}

static void rrr_logd_print (
		struct rrr_logd_data *data,
		const char *file,
		int line,
		uint8_t loglevel_translated,
		uint8_t loglevel_orig,
		uint32_t log_flags,
		const char *prefix,
		const char *message
) {
	int add_newline = 0;

	assert(message != NULL);

	if (data->add_newline && message[strlen(message) - 1] != '\n')
		add_newline = 1;

	if (data->message_only) {
		printf(add_newline ? "%s\n" : "%s", message);
		return;
	}

	if (line == 0 || file == NULL || *file == '\0') {
		line = __LINE__;
		file = __FILE__;
	}

	if (prefix == NULL || *prefix == '\0') {
		prefix = "rrr_logd";
	}

	rrr_log_print_nolock (
			file,
			line,
			loglevel_translated,
			loglevel_orig,
			prefix,
			message,
			(log_flags & RRR_MSG_LOG_F_JSON) != 0,
			add_newline
	);
}

static int rrr_logd_read_msg_callback (
		struct rrr_msg_msg **message,
		void *arg1,
		void *arg2
) {
	struct rrr_msg_msg *msg = *message;
	struct rrr_logd_callback_data *callback_data = arg2;

	(void)(arg1);

	int ret = RRR_READ_OK;

	struct rrr_array array = {0};
	char *log_message = NULL;
	char *log_prefix = NULL;
	char *log_file = NULL;
	uint8_t log_level_translated = 7;
	int log_line = 0;
	uint32_t log_flags = 0;

	if (callback_data->data->quiet)
		goto out;

	if (!MSG_IS_ARRAY(msg)) {
		RRR_MSG_0("Received non-array RRR standard message from client, this is not supported\n");
		ret = RRR_READ_SOFT_ERROR;
		goto out;
	}

	uint16_t array_version;
	if ((ret = rrr_array_message_append_to_array(&array_version, &array, msg)) != 0) {
		goto out;
	}

	if ((ret = rrr_log_helper_extract_log_fields_from_array (
			&log_file,
			&log_line,
			&log_level_translated,
			&log_flags,
			&log_prefix,
			&log_message,
			&array
	)) != 0) {
		// Returns hard or soft error
		RRR_MSG_0("Error while processing received array message\n");
		goto out;
	}

	rrr_logd_print (
			callback_data->data,
			log_file,
			log_line,
			log_level_translated,
			RRR_MSG_LOG_LEVEL_ORIG_NOT_GIVEN,
			log_flags,
			log_prefix,
			log_message
	);

	out:
	RRR_FREE_IF_NOT_NULL(log_message);
	RRR_FREE_IF_NOT_NULL(log_prefix);
	RRR_FREE_IF_NOT_NULL(log_file);
	rrr_array_clear(&array);
	return ret;
}

static int rrr_logd_read_log_callback (
		const struct rrr_msg_log *message,
		void *arg1,
		void *arg2
) {
	struct rrr_logd_callback_data *callback_data = arg2;

	(void)(arg1);

	int ret = RRR_READ_OK;

	char *log_prefix = NULL;
	char *log_message = NULL;

	if (callback_data->data->quiet)
		goto out;

	if ((ret = rrr_msg_msg_log_to_str(&log_prefix, &log_message, message)) != 0) {
		goto out;
	}

	rrr_logd_print (
			callback_data->data,
			message->file,
			message->line > INT_MAX
				? INT_MAX
				: rrr_int_from_biglength_bug_const(message->line),
			message->loglevel_translated,
			message->loglevel_orig,
			message->msg_flags,
			log_prefix,
			log_message
	);

	out:
	RRR_FREE_IF_NOT_NULL(log_prefix);
	RRR_FREE_IF_NOT_NULL(log_message);
	return ret;
}

static int rrr_logd_read_ctrl_callback (
		const struct rrr_msg *message,
		void *arg1,
		void *arg2
) {
	(void)(arg1);
	(void)(arg2);

	if (!RRR_MSG_CTRL_F_HAS(message, RRR_MSG_CTRL_F_PING)) {
		RRR_MSG_0("Received control message did not have ping flag set\n");
		return RRR_READ_SOFT_ERROR;
	}

	return 0;
}

static void rrr_logd_fd_close_callback (RRR_SOCKET_CLIENT_FD_CLOSE_CALLBACK_ARGS) {
	struct rrr_logd_data *data = arg;

	(void)(addr);
	(void)(addr_len);
	(void)(addr_string);

	if (!main_running)
		return;

	if (data->persist &&
	    create_type == RRR_SOCKET_CLIENT_COLLECTION_CREATE_TYPE_INBOUND)
		return;

	RRR_DBG_1("Received close notification for fd %i, stopping now. Was finalized is %i.\n",
		fd, was_finalized);
	main_running = 0;
}

int main (int argc, const char **argv, const char **env) {
	if (!rrr_verify_library_build_timestamp(RRR_BUILD_TIMESTAMP)) {
		fprintf(stderr, "Library build version mismatch.\n");
		exit(EXIT_FAILURE);
	}

	int ret = EXIT_SUCCESS;

	int is_child = 0;

	struct rrr_signal_handler *signal_handler = NULL;
	struct rrr_signal_handler *signal_handler_fork = NULL;
	struct cmd_data cmd = {0};
	struct rrr_logd_data data = {0};
	struct rrr_event_queue *events;
	struct rrr_socket_client_collection *clients;
	struct rrr_fork_handler *fork_handler;
	struct rrr_fork_default_exit_notification_data exit_notification_data = {
		&some_fork_has_stopped
	};
	int fd_to_wrapper[2] = {0, 0};

	if (rrr_allocator_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_final;
	}

	if (rrr_log_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_allocator;
	}

	rrr_strerror_init();

	cmd_init(&cmd, cmd_rules, argc, argv);

	if (rrr_fork_handler_new (&fork_handler) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_cmd;
	}

	signal_handler_fork = rrr_signal_handler_push(rrr_fork_signal_handler, NULL);
	signal_handler = rrr_signal_handler_push(rrr_signal_handler, NULL);
	rrr_signal_default_signal_actions_register();

	if (rrr_main_parse_cmd_arguments_and_env(&cmd, env, CMD_CONFIG_DEFAULTS) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_signal;
	}

	if (rrr_main_print_banner_help_and_version(&cmd, 3)) {
		goto out_cleanup_signal;
	}

	if (rrr_logd_parse_config(&data, &cmd) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_config;
	}

	rrr_signal_handler_set_active(RRR_SIGNALS_ACTIVE);

	if (data.unbuffered) {
		RRR_DBG_1("Setting unbuffered output\n");
		setbuf(stdout, NULL);
	}

	if (rrr_logd_socket_setup(&data) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_config;
	}

	if (rrr_event_queue_new(&events) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_socket;
	}

	if (rrr_socket_client_collection_new(&clients, events, "rrr_logd") != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_events;
	}

	struct rrr_logd_callback_data callback_data = {
		&data,
		fork_handler,
		0
	};

	rrr_socket_client_collection_event_setup (
			clients,
			NULL,
			NULL,
			NULL,
			65536,
			RRR_SOCKET_READ_METHOD_RECV | RRR_SOCKET_READ_CHECK_POLLHUP,
			NULL,
			NULL,
			rrr_logd_read_msg_callback,
			NULL,
			rrr_logd_read_log_callback,
			rrr_logd_read_ctrl_callback,
			NULL,
			&callback_data
	);

	rrr_socket_client_collection_fd_close_notify_setup (
			clients,
			rrr_logd_fd_close_callback,
			&data
	);

	if (data.receive_socket_fd > 0 && rrr_socket_client_collection_listen_fd_push (
			clients,
			data.receive_socket_fd
	) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_clients;
	}

	if (data.receive_fd > 0) {
		if (rrr_socket_check_alive (data.receive_fd, 0 /* Not silent */) != 0) {
			RRR_MSG_0("Given file descriptor %i was unusable, exiting now.\n",
				data.receive_fd);
			ret = EXIT_FAILURE;
			goto out_cleanup_clients;
		}

		if (rrr_socket_client_collection_connected_fd_push (
				clients,
				data.receive_fd,
				RRR_SOCKET_CLIENT_COLLECTION_CREATE_TYPE_PERSISTENT
		) != 0) {
			ret = EXIT_FAILURE;
			goto out_cleanup_clients;
		}
	}

	if (data.wrapper) {
		RRR_DBG_1("Starting wrapper process '%s'\n", *data.wrapper);

		for (char **arg = data.wrapper + 1; *arg; arg++) {
			RRR_DBG_1(" - arg: '%s'\n", *arg);
		}

		if (rrr_socket_pipe_blocking(fd_to_wrapper, "logd wrapper") != 0) {
			RRR_MSG_0("Failed to create pipe to wrapper\n");
			goto out_cleanup_clients;
		}

		pid_t pid = rrr_fork (
				fork_handler,
				rrr_fork_default_exit_notification,
				&exit_notification_data
		);
		if (pid < 0) {
			RRR_MSG_0("Could not fork child process: %s\n", rrr_strerror(errno));
			ret = EXIT_FAILURE;
			goto out_cleanup_pipe;
		}
		else if (pid == 0) {
			// CHILD CODE
			is_child = 1;

			RRR_DBG_1("In child wrapper\n");

			rrr_socket_close_if_set(&fd_to_wrapper[1]);

			if (rrr_socket_dup2(fd_to_wrapper[0], STDIN_FILENO) != 0) {
				RRR_MSG_0("dup2 failed in child wrapper\n");
				ret = EXIT_FAILURE;
				goto out_cleanup_pipe;
			}

			execve(data.wrapper[0], data.wrapper, (char * const *) env);

			RRR_MSG_0("Failed to execute wrapper '%s': %s\n", data.wrapper[0], rrr_strerror(errno));

			goto out_cleanup_pipe;
		}

		rrr_socket_close_if_set(&fd_to_wrapper[0]);

		RRR_DBG_1("Child wrapper started pid %i, redirecting output\n", pid);

		if ((callback_data.fd_stdout_orig = dup(STDOUT_FILENO)) < 0) {
			RRR_MSG_0("dup failed int main process\n");
			ret = EXIT_FAILURE;
			goto out_cleanup_pipe;
		}

		if (rrr_socket_dup2(fd_to_wrapper[1], STDOUT_FILENO) != 0) {
			RRR_MSG_0("dup2 failed int main process\n");
			ret = EXIT_FAILURE;
			goto out_cleanup_pipe;
		}

		RRR_DBG_1("Output now redirected\n", pid);
	}

	RRR_DBG_1("RRR log deamon starting dispatch\n");

	if (rrr_event_dispatch (
			events,
			100 * 1000, /* 100 ms */
			rrr_logd_periodic,
			&callback_data
	) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_pipe;
	}

	rrr_logd_stdout_restore(callback_data.fd_stdout_orig);

	RRR_DBG_1("RRR log deamon dispatch ended\n");

	out_cleanup_pipe:
		rrr_socket_close_if_set(&fd_to_wrapper[0]);
		rrr_socket_close_if_set(&fd_to_wrapper[1]);
	out_cleanup_clients:
		rrr_socket_client_collection_destroy(clients);
		data.receive_socket_fd = 0;
	out_cleanup_events:
		rrr_event_queue_destroy(events);
	out_cleanup_config:
		rrr_logd_cleanup_config(&data);
	out_cleanup_socket:
		if (data.receive_socket_fd > 0)
			rrr_socket_close(data.receive_socket_fd);
		rrr_socket_close_all();

	out_cleanup_signal:
		rrr_signal_handler_set_active(RRR_SIGNALS_NOT_ACTIVE);

		rrr_signal_handler_remove(signal_handler);
		rrr_signal_handler_remove(signal_handler_fork);

		if (is_child) {
			// Child forks must skip *ALL* the fork-cleanup stuff. It's possible that a
			// child which regularly calls rrr_fork_handle_sigchld_and_notify_if_needed
			// will hande a SIGCHLD before we send signals to all forks, in which case
			// it will clean up properly anyway.
			goto out_cleanup_cmd;
		}

		rrr_fork_send_sigusr1_and_wait(fork_handler);
		rrr_fork_handle_sigchld_and_notify_if_needed(fork_handler, 1);
		rrr_fork_handler_destroy (fork_handler);

	out_cleanup_cmd:
		cmd_destroy(&cmd);
		rrr_log_cleanup();

	out_cleanup_allocator:
		rrr_allocator_cleanup();

	out_final:
		return ret;
}
